dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -144,16 +144,33 @@ private boolean scheduleNextTask() {
         if (components == null) {
             return false;
         }
-
+        SettableFuture<Response> queuedResponse = components.response();
+        // If the future has been completed (most likely via cancel) the call should not be queued.
+        // There's a race where cancel may be invoked between this check and execution, but the scheduled
+        // request will be quickly cancelled in that case.
+        if (queuedResponse.isDone()) {
+            return true;
+        }
         try (CloseableSpan ignored = components.span().childSpan("Dialogue-request-scheduled")) {
-            Optional<ListenableFuture<Response>> response =
+            Optional<ListenableFuture<Response>> maybeResponse =
                     delegate.maybeExecute(components.endpoint(), components.request());
 
-            if (response.isPresent()) {
+            if (maybeResponse.isPresent()) {
+                ListenableFuture<Response> response = maybeResponse.get();
                 components.span().complete();
                 numRunningRequests.incrementAndGet();
-                response.get().addListener(numRunningRequests::decrementAndGet, DIRECT);
-                Futures.addCallback(response.get(), new ForwardAndSchedule(components.response()), DIRECT);
+                response.addListener(numRunningRequests::decrementAndGet, DIRECT);
+                Futures.addCallback(response, new ForwardAndSchedule(queuedResponse), DIRECT);
+                queuedResponse.addListener(
```

Comment:
so this is just saying if a request comes off the queue and gets sent down the delegate pipe, but the user decides to cancel the settablefuture they were holding, then we try and cancel the thing we sent down the delegate chain?

How will our lowest-level layer clients perceive this? does it just set the interrupted flag which we'd then hope something in apache would check?

Reply:
Since we use the blocking channel adapter, the thread running the request is interrupted which may result in the request stopping depending on the point in the lifecycle.